elements/pl-external-grader-variables/pl-external-grader-variables.py

```diff
@@ -1,33 +1,66 @@
 import prairielearn as pl
-import lxml.html
+import lxml
 import chevron
 
 
-PARAMS_NAME_DEFAULT = None
+def prepare(element_html, data) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    pl.check_attribs(element, ["params-name"], [])
 
+    params_name = pl.get_string_attrib(element, "params-name")
 
-def prepare(element_html, element_index, data):
-    element = lxml.html.fragment_fromstring(element_html)
-    pl.check_attribs(element, ['params-name'], [])
+    # Get frontend variables
+    frontend_variables = []
+
+    for child in element:
+        if child.tag == "pl-variable":
+            pl.check_attribs(child, ["name", "type"], ["description"])
+
+            var_dict = {
+                "name": pl.get_string_attrib(child, "name"),
+                "type": pl.get_string_attrib(child, "type"),
+            }
+
+            if pl.has_attrib(child, "description"):
+                var_dict["description"] = pl.get_string_attrib(child, "description")
+
+            frontend_variables.append(var_dict)
+
+        elif child.tag is lxml.etree.Comment:
+            continue
+
+        else:
+            raise ValueError(
+                f"Tags inside of pl-external-grader-variables must be pl-variable, not '{child.tag}'."
+            )
 
-    params_name = pl.get_string_attrib(element, 'params-name', PARAMS_NAME_DEFAULT)
-    if params_name is None:
-        raise Exception('Attribute "params-name" is not defined.')
     if params_name not in data['params']:
-        raise Exception(f"Variable name {params_name} does not exist in data['params'].")
+        data["params"][params_name] = frontend_variables
```

Comment:
Do you think we should throw an error if we don't have anything define either in the question html or the server file?  I think it could be nice to have a small sanity check, I don't think we would run in the case where we don't want *anything* from the user?

Reply:
I was actually wondering about this too. There's an example of this use case in this PR in the Pandas autograder. Basically, there's actually many instances where either the `names_from_user` or `names_for_user` are empty, and it's nice have an element fill that part of the dictionary without having to explicitly make a `server.py` just to set an empty list.

You raise an important point that not raising an error in this case could make it harder for the end user. Should we add an `empty` boolean flag as an attribute to this element? That flag would be set in the case that the user explicitly wants to declare that parameter as the empty list. That way, we can still raise an error message in the case of typos on the `params-name` attribute.

Reply:
Just added the empty flag. So to define a parameter as the empty list, you need to set the empty flag explicitly, and without this flag set, the element will throw an error if there are no backend variables defined.

Reply:
I think that checks out. I would sort of assume if there weren't any variables defined anyway then whoever is writing the question could omit the `pl-external-grader-variables` tag altogether.

Reply:
Moved the description to be the inner part of the nested `<pl-variable>` tag. To omit, the tag is left empty (but name and type can still be set).